add SHT2x sensor type

hmueller01 opened the pull request:
add SHT2xSensor class (SHT20, SHT21, SHT25)

hmueller01 committed:
add SHT2x sensor type

hmueller01 committed:
add SHT2x sensor type

winkj commented on README.md:
```diff
@@ -6,10 +6,11 @@ Repository for Sensirion humidity and temperature sensor support on Arduino
 - SHT3x-ARP (ratiometric analog voltage output)
-- SHT85
 - SHT4x
+- SHT85
```
The SHT85 is SHT3x based, so should not be part of this commit

hmueller01 replied:
Ah, ok. I just ordered it by number ... will revert this tomorrow.

winkj commented on SHTSensor.cpp:
```diff
@@ -153,6 +153,71 @@ class SHTC1Sensor : public SHTI2cSensor
+                     66,       // duration
+                     -46.85,   // a (sht_t_poly1)
+                     175.72,   // b (sht_t_poly2)
+                     65536.0,  // c (sht_t_poly3)
```
this should likely be 65535

hmueller01 replied:
Do you really want me to change this? I would prefer the datasheet value to avoid "stupid" questions ( like mine ;-) ). See PR #30

winkj replied:
For PR #30, I answered this for the SHTC1, since the case is clearer there

winkj replied:
Upon thinking about this some more, this is likely compensated by the coefficients here. So please keep 2^16 for the SHT21

winkj commented on SHTSensor.cpp:
```diff
@@ -153,6 +153,71 @@ class SHTC1Sensor : public SHTI2cSensor
+
+    // read T from SHT2x Sensor
+    if (!readFromI2c(mWire, mI2cAddress, cmd, mCmd_Size, data,
+                     EXPECTED_DATA_SIZE / 2, mDuration)) {
```
I feel using `EXPECTED_DATA_SIZE` and later `&data[3]` is a bit confusing, but don't see a better way without changing the overall code.

Could you add a strategic comment explaining how the SHT2x is different in its readout, and that we're therefore splitting the `data` into two chunks, and therefore also halving `EXPECTED_DATA_SIZE`

winkj replied:
@hmueller01

hmueller01 replied:
added an explanation that might help to understand the code

winkj commented on SHTSensor.h:
```diff
@@ -32,6 +32,18 @@
 #include <inttypes.h>
 #include <Wire.h>
 
+//#define DEBUG_SHT_SENSOR
```
Note: This could be a separate commit, but it doesn't hurt :). We could likely use this in `bool SHTI2cSensor::readSample()`, too

winkj replied:
@hmueller01 we can ignore this; for a completely independent change like adding the debug logging, I'd typically want to have this in a separate commit, but for simplicity, we can leave this one as is

hmueller01 replied:
I agree, but I needed this to implement the SHT2x code. Too bad that it wasn't there in the first place. If you like we can do a new PR to add the debug feature to the "old" code to help other developers to see what goes wrong if the sensor (!= SHT2x) does not respond ...

winkj commented:
Overall, this looks great, thanks for the submission. I don't have an SHT2x to test, but will ping a few colleagues. I've added a few minor notes

hmueller01 committed:
reverted commit

hmueller01 committed:
Merge branch 'sht2x-sensor' of https://github.com/hmueller01/arduino-sht into sht2x-sensor

hmueller01 committed:
reverted SHT order

hmueller01 commented:
FYI: I tested the code in AUTO_DETECTION on SHT21 and SHT30 ...

winkj commented on SHTSensor.cpp:
```diff
@@ -153,6 +153,71 @@ class SHTC1Sensor : public SHTI2cSensor
+
+    // convert to Temperature/Humidity
+    uint16_t val;
+    val = (data[0] << 8) + data[1];
```
upon re-reading the datasheet, we should clear the status bits here

For reference, here's what the sample code from https://sensirion.com/products/catalog/SHT21 does:

`u16sT &= ~0x0003;           // clear bits [1..0] (status bits)`

winkj replied:
@hmueller01

hmueller01 replied:
Ok, did this. Tx for the finding.

winkj commented on SHTSensor.cpp:
```diff
@@ -153,6 +153,71 @@ class SHTC1Sensor : public SHTI2cSensor
+    val = (data[0] << 8) + data[1];
+    mTemperature = mA + mB * (val / mC);
+
+    val = (data[3] << 8) + data[4];
```
same as above:

upon re-reading the datasheet, we should clear the status bits here

For reference, here's what the sample code from https://sensirion.com/products/catalog/SHT21 does:

`u16sT &= ~0x0003;           // clear bits [1..0] (status bits)`

winkj replied:
@hmueller01

hmueller01 replied:
Ok, did this. Tx for the finding.

hmueller01 commented:
@winkj Is it possible to merge this, as I'd like to see this in my production code ... Tx a lot.

winkj commented:
@hmueller01 it looks like there are still four review items open; were you planning to leave this as is?

Edit: likely my fault, the review was still open; I assumed you could see it, but chances are it was only visible to me. My apologies for that

winkj commented:
@hmueller01 at the very least, the code should clear the status bits; I can add the comments on the `EXPECTED_DATA_SIZE / 2`, and I'm happy to leave the debug code as is

hmueller01 commented:
> @hmueller01 it looks like there are still four review items open; were you planning to leave this as is?

I only see two review items and these are resolved. Or did I overlook something?

hmueller01 commented:
> @hmueller01 at the very least, the code should clear the status bits; I can add the comments on the `EXPECTED_DATA_SIZE / 2`, and I'm happy to leave the debug code as is

What do you mean with "status bits"? And comments on `EXPECTED_DATA_SIZE / 2`? I do not see a comment there ...

winkj reviewed (changes requested):
Not sure why these have not been showing up

winkj commented:
@hmueller01 the items not showing up was likely my fault, since I hadn't submitted the review. my apologies. Please let me know if the above makes sense or not.

I should be getting SHT2x inbound - hopefully arriving tomorrow - and will test & merge

hmueller01 committed:
update review findings

- added SHT2x comment to read two chunks
- check status bits
- mask status bits at calculation

winkj reviewed (dismissed):
Looks great, thanks for the updates. Will tests as soon as my SHT2x board arrives and merge

winkj commented:
Got my boards, and started testing: first problem:

```c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:216:25: warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]
     if ((data[1] & 0x02 != 0x00) || (data[4] & 0x02 != 0x20)) {
                    ~~~~~^~~~~~~
c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:216:53: warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]
     if ((data[1] & 0x02 != 0x00) || (data[4] & 0x02 != 0x20)) {
                                                ~~~~~^~~~~~~
c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:223:26: warning: suggest parentheses around '+' in operand of '&' [-Wparentheses]
     val = (data[0] << 8) + data[1] & ~0x03; // get value and clear status bits [1..0]
           ~~~~~~~~~~~~~~~^~~~~~~~~
c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:226:26: warning: suggest parentheses around '+' in operand of '&' [-Wparentheses]
     val = (data[3] << 8) + data[4] & ~0x03; // get value and clear status bits [1..0]```

winkj commented:
Second error (after fixing the parentheses from the previous comment)

```           In file included from c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:32:0:
c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp: In member function 'virtual bool SHT2xSensor::readSample()':
c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.h:38:40: error: 'class Serial_' has no member named 'printf'; did you mean 'print'?
   #define DEBUG_SHT(f,...) do { Serial.printf(PSTR(f), ##__VA_ARGS__); } while (0)```

winkj commented:
After fixing the two problems above, two separate sensor boards both fail at readFromI2c(T); will revisit at a later point in time

Issues found (see below)

winkj commented:
appears to be a timing issue with my setup: increasing duration from 66 to 100 gives me a successful results (although the status bit check still fails)

Update: found the culprit (see below)

winkj commented on SHTSensor.cpp:
```diff
@@ -153,6 +153,82 @@ class SHTC1Sensor : public SHTI2cSensor
+
+    // check status bits [1..0] (see datasheet)
+    // bit 0: not used, bit 1: measurement type (0: temperature, 1 humidity)
+    if ((data[1] & 0x02 != 0x00) || (data[4] & 0x02 != 0x20)) {
```
0x20 should likely be 0x2 here

hmueller01 replied:
Damn it. What a bad typo. Fixed.

winkj commented on SHTSensor.cpp:
```diff
@@ -153,6 +153,82 @@ class SHTC1Sensor : public SHTI2cSensor
+      // clock stretching disabled
+      : SHTI2cSensor(0x40,     // i2cAddress
+                     0xF3F5,   // i2cCommand Hi: T, Lo: RH
+                     66,       // duration
```
66 -> 85

66 is the typical duration, 85 the max

hmueller01 replied:
Hm, ok. Will increase this. All my SHT21 (5 pieces) worked with 66 ... Done.

winkj commented on SHTSensor.h:
```diff
@@ -32,6 +32,18 @@
+#ifdef DEBUG_ESP_PORT
+#define DEBUG_SHT(f,...) do { DEBUG_ESP_PORT.printf(PSTR(f), ##__VA_ARGS__); } while (0)
+#else
+#define DEBUG_SHT(f,...) do { Serial.printf(PSTR(f), ##__VA_ARGS__); } while (0)
```
Serial only has print() and println(), no printf: https://www.arduino.cc/reference/en/language/functions/communication/serial/

hmueller01 replied:
ESP8266 Arduino supports it ...
https://github.com/esp8266/Arduino/blob/5e0ed3a38bbe9a1e203e39eb4734dc236c41e57e/cores/esp8266/Print.h#L82C17-L82C17
But makes no sense using it, if other variants do not. Wasn't aware of that. At the end of the day I do not use parameter at the moment, so just using print is fine. Will fix this.

hmueller01 replied:
Ok, done with [fbadd08](https://github.com/Sensirion/arduino-sht/pull/29/commits/fbadd08e6c561f99e26c0298d77e253c6613530a).

hmueller01 commented:
> Got my boards, and started testing: first problem:
> 
> ```
>      if ((data[1] & 0x02 != 0x00) || (data[4] & 0x02 != 0x20)) {
>                     ~~~~~^~~~~~~
> c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:216:53: warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]
>      if ((data[1] & 0x02 != 0x00) || (data[4] & 0x02 != 0x20)) {
>                                                 ~~~~~^~~~~~~
> c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:223:26: warning: suggest parentheses around '+' in operand of '&' [-Wparentheses]
>      val = (data[0] << 8) + data[1] & ~0x03; // get value and clear status bits [1..0]
>            ~~~~~~~~~~~~~~~^~~~~~~~~
> c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:226:26: warning: suggest parentheses around '+' in operand of '&' [-Wparentheses]
>      val = (data[3] << 8) + data[4] & ~0x03; // get value and clear status bits [1..0]```
> ```

Sorry, I oversaw these warnings in my last update. Will fix this.

hmueller01 committed:
fixed warning: suggest parentheses around ... [-Wparentheses]

hmueller01 committed:
fixed typo in status bits check (0x20 -> 0x02)

hmueller01 committed:
increased duration from typical to the max (85)

hmueller01 committed:
replaced DEBUG output printf by print (to be compatible with official Arduino reference)

hmueller01 commented:
@winkj Please check again if all things work as expected. Tx!

winkj commented:
> @winkj Please check again if all things work as expected. Tx!

Works great now, thanks for the quick updates. I've messed up the merge, but will do a new release later this week

winkj commented:
@hmueller01 I squashed this branch into a single commit - db1c4bc738d26c89a68d0cdaa59520d831ad09b4 - and merge this into master. This is included in the v1.2.3 release that I just tagged, so that should be available via Arduino IDE for their next update cycle.

Thanks a lot for the submission and patience!

hmueller01 commented:
Thanks a lot for integrating my code into this project and for finding all the bugs I did. Really sorry that it wasn't more perfect in the first place.

Tested the released v1.2.3 on my SHT30 and SHT21 projects and to looks good. 

But I wonder that you released it as 1.2.3 and not as 1.3.0, as it is more than a small bugfix. Ok, at the end it's just a number. ;-)

hmueller01 commented:
And, documentation of 1.2.3 is a bit lame ... no hint about SHT2x

> [Version 1.2.3](https://github.com/Sensirion/arduino-sht/releases/tag/v1.2.3) [Latest](https://github.com/Sensirion/arduino-sht/releases/latest)
v1.2.3
> 
> bump version to 1.2.3

Last but not least, why is it "latest" if there is a 1.2.4 already?

winkj commented:
@hmueller01 thanks for testing the new release. I wanted to get the release out ASAP and thus was short on the description, but have now updated to add the new feature:

> [Version 1.2.3](https://github.com/Sensirion/arduino-sht/releases/tag/v1.2.3)
>
> Add support for SHT2x - Contributed by hmueller01 (https://github.com/hmueller01)

Let me know if you'd prefer to be credited differently

> Last but not least, why is it "latest" if there is a 1.2.4 already?

I kept 1.2.3 as "latest" while testing of 1.2.4. Marked as latest now as the changes in .4 seem to cause no side effects
